File: check_braces.dart
import 'dart:io';

void main(List<String> args) {
  if (args.isEmpty) return;
  final file = File(args[0]);
  final content = file.readAsStringSync();
  
  final stack = <Map<String, dynamic>>[];
  final pairs = {'(': ')', '{': '}', '[': ']'};
  
  for (var i = 0; i < content.length; i++) {
    final char = content[i];
    if (char == '(' || char == '{' || char == '[') {
      stack.add({'char': char, 'pos': i});
    } else if (char == ')' || char == '}' || char == ']') {
      if (stack.isEmpty) {
        print("Extra closing '$char' at position $i");
        printContext(content, i);
        return;
      }
      final top = stack.removeLast();
      if (pairs[top['char']] != char) {
        print("Mismatched '$char' at position $i, expected '${pairs[top['char']]}' to match '${top['char']}' from position ${top['pos']}");
        printContext(content, i);
        return;
      }
    }
  }
  
  if (stack.isNotEmpty) {
    for (final item in stack) {
      print("Unclosed '${item['char']}' from position ${item['pos']}");
      printContext(content, item['pos']);
    }
  } else {
    print("Balanced!");
  }
}

void printContext(String content, int pos) {
  final start = pos - 50 < 0 ? 0 : pos - 50;
  final end = pos + 50 > content.length ? content.length : pos + 50;
  print("Context: ${content.substring(start, end)}");
}

File: check_braces.py
import sys

def check_balance(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()
    
    stack = []
    pairs = {'(': ')', '{': '}', '[': ']'}
    for i, char in enumerate(content):
        if char in '({[':
            stack.append((char, i))
        elif char in ')}]':
            if not stack:
                print(f"Extra closing '{char}' at position {i}")
                # Print context
                start = max(0, i-20)
                end = min(len(content), i+20)
                print(f"Context: {content[start:end]}")
                return
            top, pos = stack.pop()
            if pairs[top] != char:
                print(f"Mismatched '{char}' at position {i}, expected '{pairs[top]}' to match '{top}' from position {pos}")
                # Print context
                start = max(0, i-20)
                end = min(len(content), i+20)
                print(f"Context: {content[start:end]}")
                return
    
    if stack:
        for char, pos in stack:
            print(f"Unclosed '{char}' from position {pos}")
            # Print context
            start = max(0, pos-20)
            end = min(len(content), pos+20)
            print(f"Context: {content[start:end]}")
    else:
        print("Balanced!")

if __name__ == "__main__":
    check_balance(sys.argv[1])
